modules/internal/ChapelBase.chpl

```diff
@@ -636,17 +636,73 @@ module ChapelBase {
   //
   // left and right shift on primitive types
   //
-  inline proc <<(a: int(?w), b: integral) return __primitive("<<", a, b);
-  inline proc <<(a: uint(?w), b: integral) return __primitive("<<", a, b);
 
-  inline proc >>(a: int(?w), b: integral) return __primitive(">>", a, b);
-  inline proc >>(a: uint(?w), b: integral) return __primitive(">>", a, b);
+  inline proc bitshiftChecks(a, b: integral) {
+    use HaltWrappers;
+
+    if b < 0 {
+      var msg = "Cannot bitshift " + a:string + " by " + b:string +
+                " because " + b:string + " is less than 0";
+      HaltWrappers.boundsCheckHalt(msg);
+    } else if b >= numBits(a.type) {
+      var msg = "Cannot bitshift " + a:string + " by " + b:string +
+                " because " + b:string + " is >= the bitwidth of " +
+                a.type:string;
+      HaltWrappers.boundsCheckHalt(msg);
+    }
+  }
+
+  inline proc bitshiftChecks(param a, param b: integral) {
+    param hdr = "Cannot bitshift " + a:string + " by " + b:string +
+                " because ";
+    if b < 0 {
+      param msg = hdr + b:string + " is less than 0";
+      compilerError(msg);
+    } else if b >= numBits(a.type) {
+      param msg = hdr + b:string + " is >= the bitwidth of " + a.type:string;
+      compilerError(msg);
+    }
+  }
 
-  inline proc <<(param a: int(?w), param b: integral) param return __primitive("<<", a, b);
-  inline proc <<(param a: uint(?w), param b: integral) param return __primitive("<<", a, b);
+  inline proc <<(a: int(?w), b: integral) {
+    if boundsChecking then bitshiftChecks(a, b);
+    return __primitive("<<", a:uint(w), b):int(w);
+  }
+
+  inline proc <<(a: uint(?w), b: integral) {
+    if boundsChecking then bitshiftChecks(a, b);
+    return __primitive("<<", a, b);
```

Comment:
does this primitive return `int` in all cases? regardless, I'd still cast both the argument and the return value even if it is redundant

Reply:
... and this applies to all the rest of the wrappers below